File: IOI/2007/flood/solutions/miners.cpp
#include <cstdio>
#include <cstdlib>
#include <algorithm>
using namespace std;

const int MAX=100005;

int n;
char s[MAX+1];
int dp[2][4][4][4];

void read_input() {
  scanf("%d\n", &n);
  fgets(s, MAX+1, stdin);
  for (int i = 0; i < n; i++)
    if (s[i] == 'F')
      s[i] = 1;
    else if (s[i] == 'M')
      s[i] = 2;
    else
      s[i] = 3;
}

inline int ore (int a, int b, int c) {
  int sol = 0;
  if (a == 1 || b == 1 || c == 1) sol++;
  if (a == 2 || b == 2 || c == 2) sol++;
  if (a == 3 || b == 3 || c == 3) sol++;
  return sol;
}

int main() {
  read_input();

  int now = 0, prev = 1;
  for ( int k=n-1; k>=0; --k ) {
     for ( int a=0; a<=3; ++a ) {
        for ( int b=0; b<=3; ++b ) {
           for ( int c=0; c<=3; ++c ) {
              int left = ore(s[k], s[k-1], a) + dp[prev][s[k-1]][b][c];
              int right = ore(s[k], b, c) + dp[prev][b][s[k-1]][a];
              dp[now][a][b][c] = max(left, right);
           }
        }
     }

     swap(now, prev);
  }
  printf("%d\n", dp[prev][0][0][0]);
  return 0;
}

File: IOI/2007/flood/solutions/miners.pas
program miners;

const MAX=100005;

var
   n                       : longint;
   shipment                : array[1..MAX] of longint;
   now, prev               : longint;
   dp                      : array[0..1, 0..3, 0..3, 0..3] of longint;
   k, a, b, c, left, right : longint;

procedure read_input;
var
   i  : longint;
   ch : char;
begin
   readln(n);
   for i:=1 to n do begin
      read(ch);
      case ch of
        'F' : shipment[i] := 1;
        'M' : shipment[i] := 2;
        'B' : shipment[i] := 3;
      end;
   end;
end;

function ore(a, b, c : longint) : longint;
var ret : longint;
begin
   ret := 0;
   if (a = 1) or (b = 1) or (c = 1) then ret := ret+1;
   if (a = 2) or (b = 2) or (c = 2) then ret := ret+1;
   if (a = 3) or (b = 3) or (c = 3) then ret := ret+1;
   ore := ret;
end;

function max2(a, b : longint) : longint;
begin
   if a > b then max2 := a else max2 := b;
end;

begin
   read_input;
   now := 0; prev := 1;

   for k:=n downto 1 do begin
      for a:=0 to 3 do begin
         for b:=0 to 3 do begin
            for c:=0 to 3 do begin
               left := ore(shipment[k], shipment[k-1], a) + dp[prev, shipment[k-1], b, c];
               right := ore(shipment[k], b, c) + dp[prev, b, shipment[k-1], a];
               dp[now, a, b, c] := max2(left, right);
            end;
         end;
      end;

      now := now xor 1;
      prev := prev xor 1;
   end;
   writeln(dp[prev, 0, 0, 0]);
end.
